playback/src/audio_backend/alsa.rs

```diff
@@ -194,7 +194,9 @@ impl Sink for AlsaSink {
             match open_device(&self.device, self.format) {
                 Ok((pcm, bytes_per_period)) => {
                     self.pcm = Some(pcm);
-                    self.period_buffer = Vec::with_capacity(bytes_per_period);
+                    if self.period_buffer.capacity() != bytes_per_period {
```

Comment:
Maybe a bit daft, but won't `bytes_per_period` always be the same for a given device?

Reply:
Generally yes I would think, but when open is called we start with an empty `Vec` with a capacity of zero so this is an easy way to tell if we need to actually make a new buffer or if we can reuse the current one.

Reply:
This also makes it so it wouldn't bite us in the butt if the sample rate would ever change between tracks (that would change bytes_per_period) not that Spotify does any thing but 44.1 currently AFAIK but still.

Not that the previous method would not have protected us from that also but checking against the capacity is the safest way to tell if we can reuse the same buffer. Worse case the check fails and we get a new buffer, best case we reuse the current buffer.

Reply:
Fair point, but then does it make sense to reallocate only if `bytes_per_period ` is larger that the allocated buffer?

Reply:
> Fair point, but then does it make sense to reallocate only if bytes_per_period is larger that the allocated buffer?

Yes. That would be easy enough to change `!=` to `<` but then you'd have to use `reserve` to up the capacity. And `reserve` is such that the number you feed it is what you want plus the current len of the buffer. Realistically at start the len should always be 0 but the docs say that it's possible that `reserve` can set the capacity to more than you ask for. The current buffer implementation requires that the capacity be what we want it to be. The best way to make that happen is to make a new Vec with `with_capacity` and realistically right now today  `bytes_per_period` doesn't change. The check is mostly used to tell if we're coming off an open or a stop. In the future if Spotify starts streaming things that might have a `bytes_per_period` for different tracks we could always revisit it.

Reply:
> The current buffer implementation requires that the capacity be what we want it to be.

If this is true, then we already are in trouble already. Creating a `Vec::with_capacity(x)` has the same behavior, it reserves a capacity _greater than_ or equal to `x`. You can't rely on it being exactly `x`.

Note that you'd want to use `reserve_exact(x)` (absolute) and not `reserve(x)` (relative to the current capacity). The Rust documentation is uncharacteristically unclear on this point.

Reply:
> If this is true, then we already are in trouble already. Creating a Vec::with_capacity(x) has the same behavior, it reserves a capacity greater than or equal to x. You can't rely on it being exactly x.

The buffer will never let itself grow. It uses the capacity to calculate how much it can put in itself before it calls for a write. It's such that the capacity can never grow during operation.

> Note that you'd want to use reserve_exact(x) (absolute) and not reserve(x) (relative to the current capacity). The Rust documentation is uncharacteristically unclear on this point.

The docs say that if you create a `Vec` with `with_capacity` it will be created with the given capacity. Both `reserve`'s are basically functionally identical. The intention is that both will get you "self.len() + additional".

`reserve_exact` says:

> Note that the allocator may give the collection more space than it requests. Therefore, capacity can not be relied upon to be precisely minimal. Prefer reserve if future insertions are expected.

Which seems to contradict what it ways above that:

> Reserves the minimum capacity for exactly additional more elements to be inserted in the given Vec<T>. After calling reserve_exact, capacity will be greater than or equal to self.len() + additional. Does nothing if the capacity is already sufficient.

It's a bit confusing.

Reply:
That's why I would prefer not to rely on that and for now if the capacity isn't the same just make a new buffer.

Reply:
Trust me, Google it and feel free to prove me wrong -- but not with this lacking Rust documentation.

1. If you rely on `Vec::with_capacity(period_size)` being exactly `period_size`, you can't.
2. `reserve_exact` aims to be absolute and `reserve` aims to be relative.

Rust has to rely on platform implementations so it _aims_ but cannot _guarantee_ this. If the exact capacity is important, then you should check it programmatically.

Reply:
> Trust me, Google it and feel free to prove me wrong -- but not with this lacking Rust documentation.
>
> If you rely on Vec::with_capacity(period_size) being exactly period_size, you can't.

I believe you. A `Vec` can increase in capacity. What I'm saying is that the current implementation just doesn't let that happen. It will never put more data in the `Vec` than it's capacity.

Here is `write_bytes` written more legibly:
```rust
      fn write_bytes(&mut self, data: &[u8]) -> io::Result<()> {
        let mut start_index = 0;
        while start_index < data.len() {
            let space_left = self.period_buffer.capacity() - self.period_buffer.len();
            let data_size = data.len();
            let end_index = min(space_left, data_size);
            self.period_buffer.extend_from_slice(&data[start_index..end_index]);
            start_index = end_index;
            if self.period_buffer.len() == self.period_buffer.capacity() {
                self.write_buf()?;
            }
        }

        Ok(())
    }
```

As you can see it never lets the buffer's capacity grow.

reserve_exact aims to be absolute and reserve aims to be relative.
Rust has to rely on platform implementations so it aims but cannot guarantee this. If the exact capacity is important, then you should check it programmatically.

Or you could just recreate the buffer with the desired capacity.

Reply:
You could feed that a whole song at once and it would loop though it and call `self.write_buf()?;` as many times as it took and the capacity would never grow.

Reply:
> > Rust has to rely on platform implementations so it aims but cannot guarantee this. If the exact capacity is important, then you should check it programmatically.
> 
> Or you could just recreate the buffer with the desired capacity.

No. I don't know how to explain differently.

https://github.com/librespot-org/librespot/blob/1b4d5eb7cb96a705ee861614c3afce6ba86adae4/playback/src/audio_backend/alsa.rs#L198

Does not guarantee that `self.period_buffer.capacity() == bytes_per_period` _even immediately after._

Reply:
So you can't test if the capacity is different?

Reply:
Sure that's something different. I latched onto your point that the capacity _should_ be exact and I responded with "if that's true, you're in trouble". If you _don't_ rely on this being true, then no problem.

Reply:
> Sure that's something different. I latched onto your point that the capacity _should_ be exact and I responded with "if that's true, you're in trouble". If you _don't_ rely on this being true, then no problem.

My point was that `with_capacity` *seems* to guarantee a set capacity and the `reserve`'s *don't seem to*. And when in doubt we might as well just create a new buffer instead of rely on seemly undefined behaviour.

Reply:
And on that account you are wrong. `Vec::with_capacity(x)` is like doing `Vec::new().reserve(x)`. With initial capacity being 0, `reserve` and `reserve_exact` will do the same thing. The examples Rust documentation show this, but fail to point out when capacity already is greater than 0. But I think we're going in circles, so will leave it there.

Reply:
> And on that account you are wrong.

The docs specifically say for `with_capacity`:

> It is important to note that although ***the returned vector has the capacity specified***, the vector will have a zero length. For an explanation of the difference between length and capacity, see Capacity and reallocation.

For `reserve`:

> Reserves capacity for at least additional more elements to be inserted in the given Vec<T>. The collection may reserve more space to avoid frequent reallocations. ***After calling reserve, capacity will be greater than or equal to*** self.len() + additional. Does nothing if capacity is already sufficient.

And for `reserve_exact`:

> Reserves the minimum capacity for exactly additional more elements to be inserted in the given Vec<T>. After calling reserve_exact, capacity will be greater than or equal to self.len() + additional. Does nothing if the capacity is already sufficient.
>
> Note that the allocator may give the collection more space than it requests. ***Therefore, capacity can not be relied upon to be precisely minimal.*** Prefer reserve if future insertions are expected.

Reply:
I read that to mean that with `with_capacity` you always get what you ask for with the reserves maybe not.

Reply:
And for bonus point if we actually want to avoid allocations we should keep the current period size and test against it instead of the capacity, because when capacity == len a reallocation happens.

> push and insert will never (re)allocate if the reported capacity is sufficient. push and insert will (re)allocate if len==capacity.

Reply:
> I read that to mean that with `with_capacity` you always get what you ask for but with the `reserve`s maybe not.

Like I said, dig deeper than the Rust documentation on this one. What you read into it is false. But not to make it an argument about how `with_capacity(x)` behaves, the only thing that counts is: do you or do you not rely on `period_buffer.capacity()` being exactly `x`? Or is it OK if that returns `x+1` or even `x*2`? If you do rely on it being exactly `x`, then change it so it you don't. If you already don't, then forget about it. That's the final thing I have to say about this.

Reply:
> Like I said, dig deeper than the Rust documentation on this one. What you read into it is false.

Ok. If you know you're right. I'm not trying to argue. I honestly want to know and if the rust docs are incorrect they really should be corrected. Please show me where it says different than the docs? Because the bold parts of the quotes of the docs that I posted seem to be contradicting you. Again I'm not trying to argue I honestly just want to understand it.

> The only thing that counts is: do you or do you not rely on period_buffer.capacity() being exactly x? Or is it OK if that returns x+1 or even x*2? If you do rely on it being exactly x, then change it so it you don't. If you already don't, then forget about it. That's the final thing I have to say about this.

Currently it does depend on x being exact.

Reply:
@roderickvd There Is that nay better for you?

Reply:
I wrote [this gist](https://gist.github.com/JasonLG1979/69ed65d91c64f39de4435b3720258d02) to see if `Vec` capacity can be trusted and on Linux x86_64, ARMv6 and ARMv8, it can up to 10 megabytes at least. Testing any higher would take *forever*, testing even up to 10 megabytes takes a few minutes even on my desktop, and there's no point as 10 megabytes is orders of magnitude larger than our period will ever be.

The test shows that (on Linux) `with_capacity` `reserve_exact` and `shrink_to_fit` do as you'd expect on an empty `Vec` and that you can actually `!=` test them. `reserve` however does not for values 1 - 7. You can either have 0 or more than 7. If you ask for 1 - 7 you get 8. But that's not really a concern for us since our period size will never be that small and we don't use `reserve` anyway.

I know this may be platform dependent so please compile and run the test yourself on your Mac and I invite anyone with a Windows machine to do the same.

Reply:
Well that's certainly the proof in the pudding. On my Mac I'm getting the same results, i.e. correct sizing for `with_capacity` and some failures for `reserve`.

I still think it's good to be explicit about it, and not to depend on it. There are multiple discussion on Rust issues where it's clear that memory allocation and growth strategies may - and do - change at any time.

Reply:
> I still think it's good to be explicit about it, and not to depend on it.

The current (as of before I ever touched a line of code) buffer strategy depends on capacity being dependable. I'm not sure how you could write a buffer that doesn't grow uncontrollably if you can't depend on the reported capacity?

> There are multiple discussion on Rust issues where it's clear that memory allocation and growth strategies may - and do - change at any time.

The `Vec` docs make several guarantees about capacity always being accurate but make no guarantees about allocation or growth strategies other than basically as long as you don't try to put more in then the capacity the capacity will not go up.

Reply:
If we can't 100% depend on capacity then our only choice really is a shadow variable to make sure we're actually writing the proper period size to the PCM.

Reply:
Another question:

How about turning that while loop in the buffer logic into a recessive function?

Something like this:

```rust
impl SinkAsBytes for AlsaSink {
    fn write_bytes(&mut self, data: &[u8]) -> io::Result<()> {
        self.process_data(data, 0)?;
        Ok(())
    }
}

impl AlsaSink {
    pub const NAME: &'static str = "alsa";

    fn process_data(&mut self, data: &[u8], start_index: usize) -> io::Result<()> {
        let space_left = self.period_buffer.capacity() - self.period_buffer.len();
        let data_size = data.len();
        let end_index = min(space_left, data_size);
        self.period_buffer
            .extend_from_slice(&data[start_index..end_index]);
        if self.period_buffer.len() == self.period_buffer.capacity() {
            self.write_buf()?;
        }
        if end_index < data_size {
            self.process_data(data, end_index)?;
        }
        Ok(())
    }
```

Reply:
@roderickvd as a side note if you search the code base you'll find it depend on `Vec` capacity in ***A LOT*** of places. If you don't ever want to depend on capacity you got some major refactoring to do my friend,lol!!!

Reply:
@roderickvd There I refactored the buffer to not really depend on capacity at all and hopefully not have to worry so much about the `Vec` allocation or growth strategies.

Reply:
I think the refactor fits your criteria of making the buffer implementation "more robust".

Reply:
> @roderickvd as a side note if you search the code base you'll find it depend on `Vec` capacity in _**A LOT**_ of places. If you don't ever want to depend on capacity you got some major refactoring to do my friend,lol!!!

It depends on what "depends" means, heh. I mean: will the algorithm break if the capacity is not guaranteed to be a certain `x`? Or will it yield the expected behavior, just in a suboptimal fashion or would it not even really matter?

For 99,9% cases it will just function and so in my vocabulary, it "does not depend on".

What this little deep dive taught me is that reallocation takes place when `len() == capacity()`. If I understood that right, then I am either missing something or just find it incredibly stupid. It would go on the todo list to make those `Vec::with_capacity(x)` calls do `x+1` as an optimization.

I am working now, will take a look at the different implementations hopefully later tonight.

Reply:
> What this little deep dive taught me is that reallocation takes place when `len() == capacity()`. If I understood that right, then I am either missing something or just find it incredibly stupid. It would go on the todo list to make those `Vec::with_capacity(x)` calls do `x+1` as an optimization.

From what I read I gathered the same thing. My refactor never lets `len() == capacity()` I over allocate initially to prevent that from ever happening.

I over allocate by:

`((bytes_per_period / 1024) * 1024) + 2048;`

That also makes the buffer and even amount of KiB's.

Reply:
My though (or rather hope) is that if the allocation is an even number of KiB then the underlying implementation maybe won't have to do anything weird as far as allocation. Who really knows? It doesn't hurt anything though. At the very least we're talking about a microscopic amount of memory, even when over allocating the buffer is only 45 KiB's total at F32, and it makes sure we never come anywhere near our capacity.